lib/posix/semaphore.c

```diff
@@ -148,3 +202,160 @@ int sem_wait(sem_t *semaphore)
 	(void)k_sem_take(semaphore, K_FOREVER);
 	return 0;
 }
+
+sem_t *sem_open(const char *name, int oflags, ...)
+{
+	va_list va;
+	mode_t mode;
+	unsigned int value;
+	struct nsem_obj *nsem = NULL;
+	size_t namelen;
+
+	if (name == NULL) {
+		errno = EINVAL;
+		goto error;
+	}
+
+	namelen = strlen(name);
+	if ((namelen + 1) > CONFIG_SEM_NAMELEN_MAX) {
+		errno = ENAMETOOLONG;
+		goto error;
+	}
+
+	/* Check if the named semaphore exists */
+	k_mutex_lock(&nsem_mutex, K_FOREVER);
+	nsem = find_nsem(name);
+	k_mutex_unlock(&nsem_mutex);
+
+	if ((nsem != NULL) && (oflags & O_CREAT) != 0 &&
+	    (oflags & O_EXCL) != 0) {
+		/* Named semaphore has already been opened and O_EXCL is set */
+		errno = EEXIST;
+		goto error;
+	}
+
+	if (nsem == NULL) {
+		if ((oflags & O_CREAT) == 0) {
+			errno = ENOENT;
+			goto error;
+		}
+
+		va_start(va, oflags);
+		BUILD_ASSERT(sizeof(mode_t) <= sizeof(int));
+		mode = va_arg(va, int);
+		value = va_arg(va, unsigned int);
+		va_end(va);
+
+		nsem = k_calloc(1, sizeof(struct nsem_obj));
+		if (nsem == NULL) {
+			errno = ENOSPC;
+			goto error;
+		}
+
+		/* goto `cleanup` past this point to avoid memory leak */
+
+		nsem->name = k_calloc(namelen + 1, sizeof(uint8_t));
+		if (nsem->name == NULL) {
+			errno = ENOSPC;
+			goto cleanup;
+		}
+
+		strcpy(nsem->name, name);
+		(void)atomic_set(&nsem->ref_count, 1);
+
+		if (sem_init(&nsem->sem, 0, value) != 0) {
+			errno = EINVAL;
+			goto cleanup;
+		}
+
+		k_mutex_lock(&nsem_mutex, K_FOREVER);
+		sys_slist_append(&sem_list, (sys_snode_t *)&(nsem->snode));
+		k_mutex_unlock(&nsem_mutex);
+	} else {
+		atomic_inc(&nsem->ref_count);
+	}
+
+	return &nsem->sem;
+
+cleanup:
+	cleanup_nsem(nsem);
+
+error:
+	return (sem_t *)SEM_FAILED;
+}
+
+int sem_unlink(const char *name)
+{
+	struct nsem_obj *nsem;
+	size_t namelen;
+
+	if (name == NULL) {
+		errno = EINVAL;
+		return -1;
+	}
+
+	namelen = strlen(name);
+	if ((namelen + 1)  > CONFIG_SEM_NAMELEN_MAX) {
+		errno = ENAMETOOLONG;
+		return -1;
+	}
+
+	/* Check if queue already exists */
+	k_mutex_lock(&nsem_mutex, K_FOREVER);
+	nsem = find_nsem(name);
+	if (nsem == NULL) {
+		k_mutex_unlock(&nsem_mutex);
+		errno = ENOENT;
+		return -1;
+	}
+
+	k_free(nsem->name);
```

Comment:
malloc / free would be fine to use here instead of `k_malloc()` / `k_free()`

Reply:
Are they better? I used `k_` variant as that is what the `key.c` & `mqueue.c` implementations used

Reply:
`k_` variants are fine too (probably better than malloc, to be honest)